File: ast_php/js.php
<?php
require_once __DIR__ . '/bootstrap.php';
class Printer implements LanguagePrinter
{
    private static array $currentOptions = [];

    public function startFile(): void
    {
        echo <<<'EOT'
/* @flow */
/* @generated */
/* jshint ignore:start */

/**
 * Copyright 2023-present, Lnear
 *  All rights reserved.
 */

type Node = {
  kind: string;
  start?: ?number;
  end?: ?number;
};

type OperationKind = 'query' | 'mutation' | 'subscription';


EOT;
    }

    public function endFile(): void
    {
    }

    public function startType(string $name): void
    {
        echo PHP_EOL;
        echo 'type ' . $name . ' = Node & {' . PHP_EOL;
        $kind = $name;
        if ($kind == 'GenericType') {
            $kind = 'Type';
        }
        echo '  kind: \'' . $kind . '\';' . PHP_EOL;
    }

    public function endType(string $name): void
    {
        echo '}' . PHP_EOL;
    }

    public function _jsType(string $type, bool $plural): string
    {
        if ($plural) {
            $type = 'Array<' . $type . '>';
        }
        return $type;
    }

    public function field(string $type, string $name, bool $nullable, bool $plural): void
    {
        $nullableChar = $nullable ? '?' : '';
        $jsType = $this->_jsType($type, $plural);
        echo '  ' . $name . $nullableChar . ': ' . $nullableChar . $jsType . ';' . PHP_EOL;
    }

    public function startUnion(string $name): void
    {
        echo 'type ' . $name . ' = ' . PHP_EOL;
        self::$currentOptions = [];
    }

    public function unionOption(string $type): void
    {
        self::$currentOptions[] = $type;
    }

    public function endUnion(string $name): void
    {
        echo PHP_EOL . '  | ' . join(PHP_EOL . '  | ', self::$currentOptions) . PHP_EOL;
        echo PHP_EOL;
        self::$currentOptions = [];
    }
}

File: ast/js.py

class Printer(object):
  def __init__(self):
    pass

  def start_file(self):
    print('''/* @flow */
/* @generated */
/* jshint ignore:start */

/**
 * Copyright 2023-present, Lnear
  * All rights reserved.
 */

type Node = {
  kind: string;
  start?: ?number;
  end?: ?number;
};

type OperationKind = 'query' | 'mutation' | 'subscription';''')

  def end_file(self):
    pass

  def start_type(self, name):
    print()
    print('type %s = Node & {' % name)
    kind = name
    if kind == 'GenericType':
      kind = 'Type'
    print('  kind: \'%s\';' % kind)

  def end_type(self, name):
    print('}')

  def _js_type(self, type, plural):
    if plural:
      type = 'Array<%s>' % type
    return type

  def field(self, type, name, nullable, plural):
    nullable_char = '?' if nullable else ''
    js_type = self._js_type(type, plural)
    print('  %(name)s%(nullable_char)s: %(nullable_char)s%(js_type)s;' % locals())

  def start_union(self, name):
    print(('type %s = ' % name), end=' ')
    self._current_options = []

  def union_option(self, type):
    self._current_options.append(type)

  def end_union(self, name):
    print('\n  | '.join(self._current_options))
    print()
    self._current_options = None
